flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/IntermediateResultPartitionID.java

```diff
@@ -27,29 +28,76 @@
 /**
  * Id identifying {@link IntermediateResultPartition}.
  */
-public class IntermediateResultPartitionID extends AbstractID implements ResultID {
+public class IntermediateResultPartitionID implements Comparable<IntermediateResultPartitionID>, ResultID, java.io.Serializable {
 
 	private static final long serialVersionUID = 1L;
 
+	private final IntermediateDataSetID intermediateDataSetID;
+	private final int partitionNum;
+
 	/**
-	 * Creates an new random intermediate result partition ID.
+	 * Creates an new random intermediate result partition ID for testing.
 	 */
+	@VisibleForTesting
 	public IntermediateResultPartitionID() {
-		super();
+		this.partitionNum = -1;
+		this.intermediateDataSetID = new IntermediateDataSetID(new AbstractID());
 	}
 
-	public IntermediateResultPartitionID(long lowerPart, long upperPart) {
-		super(lowerPart, upperPart);
+	/**
+	 * Creates an new intermediate result partition ID with {@link IntermediateDataSetID} and the partitionNum.
+	 */
+	public IntermediateResultPartitionID(IntermediateDataSetID intermediateDataSetID, int partitionNum) {
+		this.intermediateDataSetID = intermediateDataSetID;
+		this.partitionNum = partitionNum;
 	}
 
 	public void writeTo(ByteBuf buf) {
-		buf.writeLong(this.lowerPart);
-		buf.writeLong(this.upperPart);
+		buf.writeLong(intermediateDataSetID.getLowerPart());
```

Comment:
I think it's better to let intermediateDataSetID implement `writeTo()` and `fromByteBuf (...)` and just invokes them in IntermediateResultPartitionID.